src/guardian/guardian.service.ts

```diff
@@ -124,110 +134,249 @@ export class GuardianService implements OnModuleInit {
-        this.stakingModuleGuardService.getModulesIdsWithDuplicatedVettedUnusedKeys(
-          stakingModulesData,
-          blockData,
-        );
+      if (
+        blockData.securityVersion === 3 &&
+        !blockData.alreadyPausedDeposits &&
+        blockData.theftHappened
+      ) {
+        await this.stakingModuleGuardService.handlePauseV3(blockData);
+        return;
+      }
 
-      const stakingModulesWithoutDuplicates: StakingModuleData[] =
-        this.stakingModuleGuardService.excludeModulesWithDuplicatedKeys(
+      if (blockData.securityVersion !== 3 && blockData.theftHappened) {
+        await this.stakingModuleGuardService.handlePauseV2(
           stakingModulesData,
-          modulesIdWithDuplicateKeys,
+          blockData,
         );
+        return;
+      }
 
-      this.logger.log('Staking modules without duplicates', {
-        modulesCount: stakingModulesWithoutDuplicates.length,
-      });
-
-      await Promise.all(
-        stakingModulesData.map(async (stakingModuleData) => {
-          // stakingModulesWithoutDuplicates - modules without duplicates
-          // if found in this module it means it doesnt have duplicates
-
-          const noDuplicates = !!stakingModulesWithoutDuplicates.find(
-            (srmd) =>
-              srmd.stakingModuleId === stakingModuleData.stakingModuleId,
-          );
-
-          await this.stakingModuleGuardService.checkKeysIntersections(
-            stakingModuleData,
-            blockData,
-            noDuplicates,
-          );
-
-          this.guardianMetricsService.collectMetrics(
-            stakingModuleData,
-            blockData,
-          );
-        }),
+      this.handleKeys(stakingModulesData, blockData, lidoKeys).catch(
+        this.logger.error,
       );
 
       await this.guardianMessageService.pingMessageBroker(
         stakingModulesData.map(({ stakingModuleId }) => stakingModuleId),
         blockData,
       );
-
-      this.blockGuardService.setLastProcessedStateMeta({
-        blockHash,
-        blockNumber,
-      });
     } catch (error) {
       this.logger.error('Staking router state update error');
       this.logger.error(error);
-    } finally {
-      this.logger.log('New staking router state cycle end');
     }
   }
+
+  async collectData(
+    operatorsByModules: SROperatorListWithModule[],
+    meta: Meta,
+    lidoKeys: RegistryKey[],
+  ) {
+    const {
+      elBlockSnapshot: { blockHash, blockNumber },
+    } = meta;
+    const blockData = await this.blockGuardService.getCurrentBlockData({
+      blockHash,
+      blockNumber,
+    });
+
+    this.logger.debug?.('Current block data loaded', {
+      guardianIndex: blockData.guardianIndex,
+      blockNumber: blockNumber,
+      blockHash: blockHash,
+      securityVersion: blockData.securityVersion,
+    });
+
+    // collect some data and check keys
+    const stakingModulesData: StakingModuleData[] =
+      await this.stakingRouterService.collectStakingModuleData({
```

Comment:
I think the stakingRouterService should check the module pause status, not the blockGuardService, since blockGuardService makes requests specific to all modules, while stakingRouterService processes data for each staking module individually.